### PR TITLE
Stack check as a linear/CFG instruction

### PR DESCRIPTION
**Pull request by xclerc**
This pull request gives an explicit representation
to stack checks on the linear and CFG IRs. The
semantics is unaffected. This change is only a
preliminary step to be able to control where
the check is emitted, rather than always emit it
at the top of a function.

### CODE REVIEW

File: backend/amd64/stack_check.ml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        + (if fp then 8 else 0))
+      in
+      Misc.align sz 16
+    end else
+      stack_offset + 8
+
+let linear
+  : Linear.fundecl -> Linear.fundecl
+  = fun fundecl ->
+   (* CR mshinwell: this should be conditionalized on a specific
```
Comment (mshinwell) [inline backend/amd64/stack_check.ml:30]: I think this CR is irrelevant now and can be deleted

File: backend/amd64/stack_check.ml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        contains_nontail_calls || max_frame_size >= stack_threshold_size
+      in
+      if insert_stack_check then
+        let fun_body =
+          Linear.instr_cons
+            (Lstackcheck { max_frame_size_bytes = max_frame_size; save_registers = false})
+            [||]
+            [||]
+            ~available_before:None
+            ~available_across:None
```
Comment (mshinwell) [inline backend/amd64/stack_check.ml:52]: Let's try to set these availability values correctly (presumably based on the first instruction of `fundecl.fun_body`?)

File: backend/amd64/emit.mlp [hunk 3]
```diff
@@ -473,6 +463,36 @@ let emit_call_safety_errors () =
...
+}
+
+let stack_realloc = ref (None : stack_realloc option)
+
+let clear_stack_realloc () =
+  stack_realloc := None
+
+let emit_stack_realloc () =
+  begin match !stack_realloc with
+  | None -> ();
```
Comment (mshinwell) [inline backend/amd64/emit.mlp:480]: Spurious semicolon

File: backend/amd64/emit.mlp [hunk 4]
```diff
@@ -473,6 +463,36 @@ let emit_call_safety_errors () =
     emit_call (Cmm.global_symbol "caml_ml_array_align_error")
   end
 
+(* Stack reallocation *)
+type stack_realloc = {
+  sc_label : Label.t; (* Label of the reallocation code. *)
+  sc_return : Label.t; (* Label to return to after reallocation. *)
+  sc_max_frame_size : int; (* Size for reallocation. *)
```
Comment (mshinwell) [inline backend/amd64/emit.mlp:470]: maybe put "in bytes" here

**Review by mshinwell — CHANGES_REQUESTED** [review]
Looks good, just minor comments.

File: backend/cfg/cfg_intf.ml [hunk 5]
```diff
@@ -170,6 +170,13 @@ module S = struct
     | Pushtrap of { lbl_handler : Label.t }
     | Poptrap
     | Prologue
+    | Stack_check of
+        { max_frame_size_bytes : int;
+          save_registers : bool
```
[the remark this answers is not in the document]
Comment (xclerc) [reply backend/cfg/cfg_intf.ml]: Indeed, `save_registers` is simply "not the first instruction".

**Review by mshinwell — APPROVED** [review]

**Pull request merged**